### PR TITLE
126 refactor kafka clients and dependent services

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
Refactor kafka clients to use new spdlog macros to include filename and file line. Also refactor kafka_clients executable and test to use streets_service_base library to parse configuration file instead of custom configuration parsing logic. Removed custom configuration parsing logic.

Also added refactor of dependent services (scheduling_service, message_services).

## Related Issue
#126 

## Motivation and Context

Refactor all CARMA-Streets services to use streets_configuration's configuration parsing/retrieval code.
## How Has This Been Tested?

1. From `carma-streets/` directory call:
2. `docker build -t kafka_clients -f kafka-clients/Dockerfile .`
3. `docker build -t message_services -f message_serivces/Dockerfile .`
4. `docker build -t scheduling_service -f scheduling_service/Dockerfile .`
5. Replace image name and tag in `docker-compose.yml` with the newly built images.
6. Deploy system using `docker-compose up -d` and use `docker logs <service_name>` to see the logs of each new service.
7. Note filename and file line additions to log statements.

Also can `docker exec -it <service_name> bash` into containers and run unit test to confirm results and coverage ( Coverage available on sonar cloud and unit test pass fail results are available on circleci)

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x ] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: kafka_clients/CMakeLists.txt [hunk 1]
```diff
@@ -1,35 +1,43 @@
 cmake_minimum_required(VERSION 3.10.2)
 project(kafka_clients)
 
-link_directories("/usr/lib"  "/usr/local/lib" )
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
Comment (dan-du-car) [inline kafka_clients/CMakeLists.txt:4]: Why do we add this?

File: kafka_clients/CMakeLists.txt [hunk 2]
```diff
@@ -1,35 +1,43 @@
...
 
-find_package(Boost COMPONENTS thread filesystem system REQUIRED)
+
+
+find_package(Boost 1.65.1 COMPONENTS system filesystem thread REQUIRED)
```
Comment (dan-du-car) [inline kafka_clients/CMakeLists.txt:8]: Why do we restrict to 1.65.1 version?

File: kafka_clients/CMakeLists.txt [hunk 3]
```diff
@@ -51,4 +59,23 @@ add_executable(${BINARY} ${TEST_SOURCES}
...
+                            ${PROJECT_NAME}_lib 
+                            gtest)
+target_include_directories(${BINARY}
+                           PUBLIC
+                           $<INSTALL_INTERFACE:include>
+                           $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+                           PRIVATE
+                           ${CMAKE_CURRENT_SOURCE_DIR}/src})
+# get_cmake_property(_variableNames VARIABLES)
+# list (SORT _variableNames)
```
Comment (dan-du-car) [inline kafka_clients/CMakeLists.txt:78]: remove comment out code

File: scheduling_service/Dockerfile [hunk 4]
```diff
@@ -45,18 +45,28 @@ WORKDIR /home/carma-streets/ext/rapidjson/build
...
+
 # Install kafka-clients
 RUN echo " ------> Install kafka-clients..."
 COPY ./kafka_clients/ /home/carma-streets/kafka_clients
 WORKDIR  /home/carma-streets/kafka_clients
 RUN mkdir build
 WORKDIR  /home/carma-streets/kafka_clients/build
 RUN cmake -DCMAKE_BUILD_TYPE="Debug" ..
-RUN make
+RUN make VERBOSE=true
```
Comment (dan-du-car) [inline scheduling_service/Dockerfile:66]: Why need VERBOSE?

File: kafka_clients/CMakeLists.txt [hunk 5]
```diff
@@ -1,35 +1,43 @@
...
+find_package(Boost 1.65.1 COMPONENTS system filesystem thread REQUIRED)
 find_package(spdlog REQUIRED)
+find_package(streets_service_base_lib REQUIRED)
 
 add_library(${PROJECT_NAME}_lib STATIC 
             src/kafka_producer_worker.cpp 
             src/kafka_consumer_worker.cpp            
             src/kafka_client.cpp )
 
-include_directories( ${PROJECT_NAME}_lib PUBLIC ${PROJECT_SOURCE_DIR}/include
```
Comment (dan-du-car) [inline kafka_clients/CMakeLists.txt:14]: Where do we include the header files?

File: kafka_clients/CMakeLists.txt (hunk 1 above)
Comment (paulbourelly999) [reply kafka_clients/CMakeLists.txt:4]: This flag is for **Position Independent Code**. This produces libraries and executables that are position independent in that they can be moved and the paths still resolve correctly I think. I may have added this to attempt to resolve some linker errors but I think it is also correct to include.

File: kafka_clients/CMakeLists.txt (hunk 2 above)
Comment (paulbourelly999) [reply kafka_clients/CMakeLists.txt:8]: This is probably not necessary. I can remove this.

File: kafka_clients/CMakeLists.txt (hunk 3 above)
Comment (paulbourelly999) [reply kafka_clients/CMakeLists.txt:78]: I will remove this

File: scheduling_service/Dockerfile (hunk 4 above)
Comment (paulbourelly999) [reply scheduling_service/Dockerfile:66]: I will remove this

File: kafka_clients/CMakeLists.txt (hunk 5 above)
Comment (paulbourelly999) [reply kafka_clients/CMakeLists.txt:14]: at the target_include_directories

File: message_services/CMakeLists.txt [hunk 6]
```diff
@@ -6,23 +6,25 @@ project(message_services LANGUAGES CXX)
...
-find_package(Boost COMPONENTS thread filesystem system REQUIRED)
+find_package(Boost COMPONENTS thread system REQUIRED)
 find_package(spdlog REQUIRED)
-include_directories("/usr/local/include")
 
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include ${CMAKE_CURRENT_SOURCE_DIR}/src/models)
 
 
-add_library(${PROJECT_NAME}_lib 
+add_library(${PROJECT_NAME}_lib STATIC
```
Comment (paulbourelly999) [inline message_services/CMakeLists.txt:27]: I dont think i need static here. Will remove

File: scheduling_service/Dockerfile [hunk 7]
```diff
@@ -71,7 +81,7 @@ WORKDIR  /home/carma-streets/scheduling_service
 RUN mkdir build
 WORKDIR  /home/carma-streets/scheduling_service/build
 RUN cmake -DCMAKE_BUILD_TYPE="Debug" ..
-RUN make
+RUN make VERBOSE=true
```
Comment (paulbourelly999) [inline scheduling_service/Dockerfile:84]: Will remove this as well

File: scheduling_service/src/main.cpp [hunk 8]
```diff
@@ -550,35 +516,33 @@ void call_consumer_thread()
...
-    std::string topic = client->get_value_by_doc(doc_json, "PRODUCER_TOPIC");
+    std::string bootstrap_server =  streets_service::streets_configuration::get_string_config("bootstrap_server");
+    std::string topic = streets_service::streets_configuration::get_string_config("producer_topic");
     auto producer_worker  = client->create_producer(bootstrap_server, topic);
      // Holds timestamp for last schedule sent
     double last_schedule;
 
     // Create logger
-    if ( config.isScheduleLoggerEnabled() ) {
+    if ( streets_service::streets_configuration::get_boolean_config("enable_schedule_logging") ) {
```
Comment (dan-du-car) [inline scheduling_service/src/main.cpp:526]: What if this enable_schedule_logging is not passed in as boolean?

File: scheduling_service/src/main.cpp [hunk 9]
```diff
@@ -493,24 +461,22 @@ void call_consumer_thread()
 {
   
     auto client = std::make_shared<kafka_clients::kafka_client>();
-    std::string file_name= "../manifest.json";
-    rapidjson::Document doc_json = client->read_json_file(file_name); 
-    std::string bootstrap_server = client->get_value_by_doc(doc_json, "BOOTSTRAP_SERVER");
-    std::string group_id = client->get_value_by_doc(doc_json, "GROUP_ID");
-    std::string topic = client->get_value_by_doc(doc_json, "CONSUMER_TOPIC");
+    std::string bootstrap_server = streets_service::streets_configuration::get_string_config("bootstrap_server");
+    std::string group_id = streets_service::streets_configuration::get_string_config("group_id");
```
Comment (dan-du-car) [inline scheduling_service/src/main.cpp:465]: What if group id / bootstrap server etc  is not a string?

File: scheduling_service/src/main.cpp [hunk 10]
```diff
@@ -375,7 +343,7 @@ rapidjson::Value scheduling_func(unordered_map<string, vehicle> list_veh, Docume
                 string vehicle_id1 = schedule.get_vehicleIdList()[vehicle_index1];
                 string link_id1 = list_veh[vehicle_id1].get_linkID();
 
-                double st = max(schedule.get_estList()[vehicle_index1], schedule.get_timestamp() + config.get_schedulingDelta());
+                double st = max(schedule.get_estList()[vehicle_index1], schedule.get_timestamp() + streets_service::streets_configuration::get_double_config("scheduling_delta"));
```
Comment (dan-du-car) [inline scheduling_service/src/main.cpp:346]: The same as below

File: scheduling_service/src/main.cpp (hunk 8 above)
Comment (paulbourelly999) [reply scheduling_service/src/main.cpp:526]: an exception will be thrown

File: scheduling_service/src/main.cpp (hunk 10 above)
Comment (paulbourelly999) [reply scheduling_service/src/main.cpp:346]: an exception will be thrown

File: scheduling_service/src/main.cpp (hunk 9 above)
Comment (paulbourelly999) [reply scheduling_service/src/main.cpp:465]: an exception will be thrown

File: message_services/CMakeLists.txt (hunk 6 above)
Comment (paulbourelly999) [reply message_services/CMakeLists.txt]: Nvm static is necessary here, without it we run into the following exception(https://app.circleci.com/pipelines/github/usdot-fhwa-stol/carma-streets/660/workflows/464514d0-f154-48e6-9008-c7f721a628d1/jobs/1436?invite=true#step-104-509)
`/usr/bin/ld: /usr/local/lib/libspdlog.a(spdlog.cpp.o): relocation R_X86_64_TPOFF32 against `_ZGVZN6spdlog7details2os9thread_idEvE3tid' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
CMakeFiles/message_services_lib.dir/build.make:443: recipe for target 'libmessage_services_lib.so' failed
make[2]: *** [libmessage_services_lib.so] Error 1
CMakeFiles/Makefile2:297: recipe for target 'CMakeFiles/message_services_lib.dir/all' failed
make[1]: *** [CMakeFiles/message_services_lib.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
[  8%] Built target gtest
[ 12%] Linking CXX shared library libmessage_services_lib.so
/usr/bin/ld: /usr/local/lib/libspdlog.a(spdlog.cpp.o): relocation R_X86_64_TPOFF32 against `_ZGVZN6spdlog7details2os9thread_idEvE3tid' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
CMakeFiles/message_services_lib.dir/build.make:443: recipe for target 'libmessage_services_lib.so' failed
make[2]: *** [libmessage_services_lib.so] Error 1
CMakeFiles/Makefile2:297: recipe for target 'CMakeFiles/message_services_lib.dir/all' failed
make[1]: *** [CMakeFiles/message_services_lib.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2`

**Review by dan-du-car — APPROVED** [review]

**Pull request merged**